### PR TITLE
Don't allow instance ipv{n}.address to be same as managed parent network

### LINKED ISSUE #9089: Don't allow instance bridged NIC ipv{n}.address setting to be set to that of the managed parent network
Currently setting an instance `bridged` NIC's `ipv{n}.address` to that of the parent managed network's address is allowed (although it doesn't actually work, as dnsmasq will ignore it).

But its not a good user experience, so we should prevent it.

Same goes for `ovn` networks.

### PR DESCRIPTION
**Pull request by masnax**
Closes #9089 

Signed-off-by: Max Asnaashari <max.asnaashari@canonical.com>

### CODE REVIEW

File: lxd/device/nic_bridged.go [hunk 1]
```diff
@@ -113,6 +113,16 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
 			if dhcpv4Subnet != nil && !dhcpalloc.DHCPValidIP(dhcpv4Subnet, nil, net.ParseIP(d.config["ipv4.address"])) {
 				return fmt.Errorf("Device IP address %q not within network %q subnet", d.config["ipv4.address"], n.Name())
 			}
+
+			ip, _, err := net.ParseCIDR(netConfig["ipv4.address"])
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:117]: What happens when `ipv4.address` isn't set?

File: lxd/device/nic_bridged.go [hunk 2]
```diff
@@ -113,6 +113,16 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
...
+			if err != nil {
+				return errors.Wrap(err, "unable to parse parent network address")
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:119]: Need to capitalize that error message

File: lxd/device/nic_bridged.go (hunk 1 above)
Comment (stgraber) [reply lxd/device/nic_bridged.go:117]: It can be empty or set to `none` on a network.

File: lxd/device/nic_bridged.go [hunk 3]
```diff
@@ -113,6 +113,16 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
...
+			}
+
+			// IP should not be the same as the parent managed network address.
+			if ip.Equal(net.ParseIP(d.config["ipv4.address"])) {
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:123]: What happens if the device doesn't have an `ipv4.address` set?

File: lxd/device/nic_bridged.go [hunk 4]
```diff
@@ -129,6 +139,16 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
 			if dhcpv6Subnet != nil && !dhcpalloc.DHCPValidIP(dhcpv6Subnet, nil, net.ParseIP(d.config["ipv6.address"])) {
 				return fmt.Errorf("Device IP address %q not within network %q subnet", d.config["ipv6.address"], n.Name())
 			}
+
+			ip, _, err := net.ParseCIDR(netConfig["ipv6.address"])
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:143]: Same question around handling of `ipv6.address` being empty or being set to `none`.

File: lxd/device/nic_bridged.go [hunk 5]
```diff
@@ -129,6 +139,16 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
...
+			if err != nil {
+				return errors.Wrap(err, "unable to parse parent network address")
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:145]: Capitalize

File: lxd/device/nic_bridged.go [hunk 6]
```diff
@@ -129,6 +139,16 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
...
+			}
+
+			// IP should not be the same as the parent managed network address.
+			if ip.Equal(net.ParseIP(d.config["ipv6.address"])) {
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:149]: Same question around `ipv6.address` not being set on the NIC.

File: lxd/device/nic_ovn.go [hunk 7]
```diff
@@ -148,6 +148,16 @@ func (d *nicOVN) validateConfig(instConf instance.ConfigReader) error {
 		if !dhcpalloc.DHCPValidIP(subnet, nil, net.ParseIP(d.config["ipv4.address"])) {
 			return fmt.Errorf("Device IP address %q not within network %q subnet", d.config["ipv4.address"], d.config["network"])
 		}
+
```
Comment (stgraber) [inline lxd/device/nic_ovn.go:151]: Repeat of the questions from the bridged implementation :)

File: lxd/device/nic_bridged.go (hunk 1 above)
Comment (masnax) [reply lxd/device/nic_bridged.go:117]: Ah yeah, there's a case where that's allowed with `security.ipv4_filtering` set to true, which I missed.

File: lxd/device/nic_bridged.go (hunk 3 above)
Comment (masnax) [reply lxd/device/nic_bridged.go:123]: It's not captured in the diff, but this whole block is inside a check for just that.

File: lxd/device/nic_bridged.go [hunk 8]
```diff
@@ -99,7 +99,7 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
 
 		netConfig := n.Config()
 
-		if d.config["ipv4.address"] != "" {
+		if shared.StringInSlice(d.config["ipv4.address"], []string{"", "none"}) {
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:102]: `none` isn't a valid value for a `ipv4.address` when set on a NIC.
It's a valid value when set on a network.

File: lxd/device/nic_bridged.go [hunk 9]
```diff
@@ -113,9 +113,23 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
 			if dhcpv4Subnet != nil && !dhcpalloc.DHCPValidIP(dhcpv4Subnet, nil, net.ParseIP(d.config["ipv4.address"])) {
 				return fmt.Errorf("Device IP address %q not within network %q subnet", d.config["ipv4.address"], n.Name())
 			}
+			parentAddress, ok := netConfig["ipv4.address"]
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:116]: Line break before this

File: lxd/device/nic_bridged.go [hunk 10]
```diff
@@ -113,9 +113,23 @@ func (d *nicBridged) validateConfig(instConf instance.ConfigReader) error {
...
+			if !ok {
```
Comment (stgraber) [inline lxd/device/nic_bridged.go:117]: That should skip if it's set but is empty or set to `none`.

File: lxd/device/nic_ovn.go (hunk 7 above)
Comment (masnax) [reply lxd/device/nic_ovn.go:151]: So for the ovn implementation, there's no special flags that allow us to continue on when `network.DHCPv4Subnet()` is nil, so we'll error out before my changes.

**Review by stgraber — APPROVED** [review]

**Pull request merged**